content/v2.0/actions/parameters.md

```diff
@@ -0,0 +1,350 @@
+
+This validation serves several purposes, including allowlisting (ensuring that only allowable params are extracted from a request) and coercion (converting string parameters to boolean, integer, time and other types).
+
+Let's take a look at a books index action that accepts two parameters, `page` and `per_page`.
+
+```ruby
+# app/actions/books/index.rb
+
+# frozen_string_literal: true
+
+module Bookshelf
+  module Actions
+    module Books
+      class Index < Bookshelf::Action
+        params do
+          optional(:page).value(:integer)
+          optional(:per_page).value(:integer)
+        end
+
+        def handle(request, response)
+          request.params[:page]
+          request.params[:per_page]
+        end
+      end
+    end
+  end
+end
+```
+
+The schema in the params block specifies the following:
+
+- `page` and `per_page` are both optional parameters
+- if `page` is present, it must be an integer
+- if `per_page` is present, it must be an integer
+
+With this schema in place, a request with a query string of `/books?page=1&per_page=10` will result in:
+
+```ruby
+request.params[:page] # => 1
+request.params[:per_page] # => 10
+```
+
+Note that, due to the `params` block, 1 and 10 have been coerced to integer values (with this they would otherwise be strings).
+
+Additional rules can be added to apply further contraints. The following params block specifies that, when present, `page` and `per_page` be greater than or equal to 1, and also that `per_page` be less than or equal to 100.
+
+```ruby
+params do
+  optional(:page).value(:integer, gte?: 1)
+  optional(:per_page).value(:integer, gte?: 1, lteq?: 100)
+end
+```
+
+Importantly, now that our params block is doing more than just type coercion, we need to explicitly check for and handle our parameters being invalid.
+
+The `#valid?` method on the params allows the action to check the parameters in order halt and return a `422 Unprocessable` response.
+
+```ruby
+# app/actions/books/index.rb
+
+# frozen_string_literal: true
+
+module Bookshelf
+  module Actions
+    module Books
+      class Index < Bookshelf::Action
+        params do
+          optional(:page).value(:integer, gte?: 1)
+          optional(:per_page).value(:integer, gte?: 1, lteq?: 100)
+        end
+
+        def handle(request, response)
+          halt 422 unless request.params.valid?
+
+          # At this point, we know the params are valid
+          request.params[:page]
+          request.params[:per_page]
+        end
+      end
+    end
+  end
+end
+```
+
+Here's a further example, this time for an action to create a user.
+
+```ruby
+# app/actions/users/create.rb
+
+# frozen_string_literal: true
+
+module Bookshelf
+  module Actions
+    module Users
+      class Create < Bookshelf::Action
+        params do
+          required(:email).filled(:string)
+          required(:password).filled(:string)
+
+          required(:address).schema do
```

Comment:
```suggestion
          required(:address).hash do
```

plain `schema` doesn't specify the type to be `:hash`, which is what `hash` macro does.